File: 402-remove-k-digits/remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        if(num.length()==k){
            return "0";
        }
        Stack<Character> st=new Stack<>();
        for(int i=0;i<num.length();i++){
            while(!st.isEmpty() && st.peek()>num.charAt(i) && k>0){
                st.pop();
                k--;
            }
            st.push(num.charAt(i));
        }
        StringBuilder sb=new StringBuilder();
        while(!st.isEmpty() && k-->0){
            st.pop();
        }
        while(!st.isEmpty()){
            sb.append(st.pop());
        }
        sb.reverse();
        while(sb.length()>0 && sb.charAt(0)=='0'){
            sb.deleteCharAt(0);
        }
        return sb.length()==0 ?"0" :sb.toString();
    }
}

File: 402-remove-k-digits/remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        if(num.length()==k){
            return "0";
        }
        stack<char> st;
        string s;
        int n=num.length();
        for(int i=0;i<n;i++){
            while(!st.empty() && st.top()>num[i] && k>0){
                st.pop();
                k--;
            }
            st.push(num[i]);
        }
        while(!st.empty() && k>0){
            st.pop();
            k--;
        }
        while(!st.empty()){
            s+=st.top();
            st.pop();
        }
        reverse(s.begin(),s.end());
        while(s.length()>0 && s[0]=='0'){
            s.erase(0,1);
        }
        return s.length()==0 ? "0" : s;
    }
};